libflux/flux-core/src/formatter/mod.rs

```diff
@@ -304,28 +344,47 @@ impl Formatter {
             ast::MonoType::Function(fun) => self.format_function_type(fun),
         }
     }
+
     fn format_function_type(&mut self, n: &ast::FunctionType) {
-        let multiline = n.parameters.len() > 4 || n.base.is_multiline();
+        let mut multiline = n.parameters.len() > 4;
         self.format_comments(&n.base.comments);
         self.write_rune('(');
+
+        let mut temp_formatters: Vec<Formatter> = Vec::new();
+
+        for p in &n.parameters {
+            let mut temp = self.create_temp_formatter();
+            temp.format_parameter_type(p);
+            // if any child node contains newlines, then that child and the parent node will be
+            // multiline as well
+            if temp.builder.contains("\n") {
+                multiline = true;
+            }
+            temp_formatters.push(temp);
+        }
+
         if multiline {
             self.write_rune('\n');
             self.indent();
             self.write_indent();
         }
+
         let sep = match multiline {
             true => ",\n",
             false => ", ",
         };
-        for (i, p) in (&n.parameters).iter().enumerate() {
-            if i != 0 {
+
+        for i in 0..n.parameters.len() {
+            temp_formatters[i].reindent(self.indentation);
+            self.write_string(&temp_formatters[i].builder);
+            if i < n.parameters.len() - 1 {
                 self.write_string(sep);
                 if multiline {
                     self.write_indent();
```

Comment:
Regarding indents appearing on blank lines, it seems like we could avoid this if we indented before writing each item in a list rather than after. But thinking it over I think there could be a lot of ripple effects doing it this way and it's not really a big deal.

Reply:
Yeah, I think that not indenting blank lines would be better addressed as a follow-up PR just to try and limit what's addressed here.